### PR TITLE
feat(nexus-shutdown): adding changes for nexus shutdown api call

### PR DESCRIPTION
**Pull request by abhilashshetty04**
This PR includes implementation to call io-engine nexus shutdown api from control plane.

Signed-off-by: Abhilash Shetty <abhilash.shetty@datacore.com>

### CODE REVIEW

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/agents/src/bin/core/controller/wrapper.rs [hunk 1]
```diff
@@ -1149,9 +1148,15 @@ impl ClientOps for Arc<tokio::sync::RwLock<NodeWrapper>> {
...
-        // todo: add implementation
-        Ok(())
+    async fn shutdown_nexus(&self, request: &ShutdownNexus) -> Result<(), SvcError> {
+        let dataplane = self.grpc_client_locked(request.id()).await?;
+        let result = dataplane.shutdown_nexus(request).await;
+        let mut ctx = dataplane.reconnect(GETS_TIMEOUT).await?;
+        self.update_nexus_states(ctx.deref_mut()).await?;
+        match result {
+            Ok(_) => Ok(()),
+            Err(error) => Err(error),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/wrapper.rs:1159]: need to consider idempotence here, what happens if the nexus is already shutdown; for now we could check if the nexus state is shutdown but that doesn't cover the case when shutdown is in progress but not complete.
@mtzaurus

File: control-plane/agents/src/bin/core/controller/wrapper.rs [hunk 2]
```diff
@@ -751,8 +750,8 @@ pub(crate) trait ClientOps {
     async fn remove_child(&self, request: &RemoveNexusChild) -> Result<(), SvcError>;
     /// Fault a child from its parent nexus via gRPC.
     async fn fault_child(&self, request: &FaultNexusChild) -> Result<(), SvcError>;
-    /// Shutdown an existing nexus.
-    async fn shutdown_nexus(&self, _: &ShutdownNexus) -> Result<(), SvcError>;
+    /// Shutdown a nexus via gRPC
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/wrapper.rs]: missing `.`

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 1 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs:1159]: I need to see dataplane implementation to handle those scenarios. Will do that.

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 2 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: Will add

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 1 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs:1159]: As discussed, We will have a separate PR to handle idempotency for this. We need to decide idempotency scenarios.

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 2 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: Done

**Comment by abhilashshetty04** [comment]
bors try

**Review by tiagolobocastro — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**